### PR TITLE
[APIDiff] Improve API digester integration & polish of experimental-api-diff

### PR DESCRIPTION
**Pull request by owenv**
Improve the swift-api-digester integration of the experimental-api-diff command by taking advantage of recent compiler improvements, and fix various bugs.

### Motivation:

Make the experimental-api-diff command suitable for use in a production environment on macOS

### Modifications:

- Parse serialized diagnostic output from the API digester to differentiate breaking changes from other diagnostics
- Improve output formatting
- Generate and compare baselines on a per-module basis
- Only include modules in the analysis if they're vended by library products
- Remove some option-parsing hacks that aren't required in newer toolchains
- Improve test coverage for the feature

### Result:

After this change, experimental-api-diff should be much easier to use and mostly production-ready on macOS. Further changes will be required to get it working well on Linux & Windows because of SDK layout differences.

After this change, the command will only work with a swift-api-digester from the 5/5/2021 nightly toolchain or later. The tests already needed to be skipped on the self-hosted bots since Xcode toolchains won't include the tool until Swift 5.5, so I don't expect this to be an issue.

### CODE REVIEW

**Comment by owenv** [comment]
@swift-ci please smoke test

**Comment by owenv** [comment]
https://github.com/apple/swift-tools-support-core/pull/213
@swift-ci please smoke test

**Comment by owenv** [comment]
Hmm, looks like the self-hosted bots don't support cross-repo testing

**Comment by owenv** [comment]
apple/swift-tools-support-core#213
@swift-ci please smoke test

**Comment by owenv** [comment]
@swift-ci please smoke test

File: Sources/Build/BuildPlan.swift [hunk 1]
```diff
@@ -1919,14 +1919,7 @@ public class BuildPlan {
         // Add search paths from the system library targets.
         for target in graph.reachableTargets {
             if let systemLib = target.underlyingTarget as? SystemLibraryTarget {
-                for flag in self.pkgConfig(for: systemLib).cFlags {
```
Comment (elsh) [inline Sources/Build/BuildPlan.swift:1922]: Are these changes unnecessary now?

File: Sources/Commands/APIDigester.swift [hunk 2]
```diff
@@ -187,6 +206,21 @@ public struct SwiftAPIDigester {
...
+extension SwiftAPIDigester {
+    /// The result of comparing a module's API to a provided baseline.
+    public struct ComparisonResult {
+        /// Breaking changes made to the API since the baseline was generated.
+        var apiBreakingChanges: [SerializedDiagnostics.Diagnostic]
+        /// Other diagnostics emitted while comparing the current API to the baseline.
+        var otherDiagnostics: [SerializedDiagnostics.Diagnostic]
+
+        /// `true` if the comparison succeeded and no breaking changes were found, otherwise `false`.
+        var isSuccessful: Bool {
```
Comment (elsh) [inline Sources/Commands/APIDigester.swift:218]: Maybe rename 'hasNoAPIBreakingChanges' (or 'hasAPIBreakingChanges' that returns the opposite value)? ComparisonResult.isSuccessful doesn't necessarily convey what's successful.

File: Sources/Commands/SwiftPackageTool.swift [hunk 3]
```diff
@@ -319,14 +319,55 @@ extension SwiftPackageTool {
                 apiDigesterTool: apiDigesterTool,
                 diags: swiftTool.diagnostics
             )
-            let baselineSDKJSON = try baselineDumper.dumpBaselineSDKJSON()
+            let baselineDir = try baselineDumper.emitAPIBaseline()
+
+            var succeeded = true
+            for module in try buildOp.getPackageGraph().apiDigesterModules {
```
Comment (elsh) [inline Sources/Commands/SwiftPackageTool.swift:325]: Could you provide a reason for doing this per module? This triggers a process launch for each, which can cause issues such as a (noticeable) perf regression even when there's a minor non-api breaking change.

**Review by elsh — CHANGES_REQUESTED** [review]

File: Sources/Build/BuildPlan.swift (hunk 1 above)
Comment (owenv) [reply Sources/Build/BuildPlan.swift:1922]: Yeah, as of https://github.com/apple/swift/pull/37206 this is unnecessary because the API digester and compiler are using the same option definitions

File: Sources/Commands/APIDigester.swift (hunk 2 above)
Comment (owenv) [reply Sources/Commands/APIDigester.swift]: sounds good to me, done!

File: Sources/Commands/SwiftPackageTool.swift (hunk 3 above)
Comment (owenv) [reply Sources/Commands/SwiftPackageTool.swift:325]: Right now the API digester assumes the baseline was generated from the same set of modules it's loading, so generating a baseline from multiple and then comparing that to one module at a time returns a bunch of false positives. I also didn't want to load every module at once because currently the source module for each breaking change isn't serialized, so there isn't a reliable way to sort the changes by product/target. Future compiler improvements might let us rollback this part of the change

Comment (elsh) [reply Sources/Commands/SwiftPackageTool.swift:325]: Seems like making changes in the detection logic itself in the frontend should be the way to go (cc @nkcsgexi). Having to iterate over each module and launching the process for each won't scale. Is there at least a way to find depending modules that are affected by a diff so only those can be iterated over?

Comment (nkcsgexi) [reply Sources/Commands/SwiftPackageTool.swift:325]: Eliminating false positives at all from the API digester tool is a very challenging and long-term task. Why do you think it won't scale for iterating over each module? Build system/swift-driver iterates each modules and build them.

Comment (elsh) [reply Sources/Commands/SwiftPackageTool.swift:325]: @nkcsgexi  Build system performs several things including dependency scanning and caching/cache lookups so incremental builds are faster. The api-digester doesn't do any of that and this change launches a frontend process for each module even when there's a very minor change at the top module.

**Review by elsh — CHANGES_REQUESTED** [review]
@abertelrud @tomerd  Could you weigh in? See the discussions re: perf.

**Comment by elsh** [comment]
@owenv Since changing the detection logic itself is supposedly a difficult task, could you add support to allow an input module list (e.g. --module-list) and iterate over that specified list if provided? This way users have an option to run the api digester with a few specific modules that they know are likely affected by their changes vs unnecessarily having to compare with all modules. 

Also would be great if you could add an explanation for the experimental-api-diff subcommand and its default behavior (iterating over all modules).

**Comment by owenv** [comment]
@elsh Sure, I've added support for filtering the comparison to a subset of the package. I chose `--product` and `--target` because I thought those would be more intuitive when working on a package, but I'm open to switching to `--module` or `--module-list` if you think that would be better. I've also updated the help text and parallelized the api-digester invocations.

**Comment by owenv** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
> @elsh Sure, I've added support for filtering the comparison to a subset of the package. I chose `--product` and `--target` because I thought those would be more intuitive when working on a package, but I'm open to switching to `--module` or `--module-list` if you think that would be better. I've also updated the help text and parallelized the api-digester invocations.

Target and product seem reasonable to me, since that's the terminology used for packages.

File: Sources/Commands/SwiftPackageTool.swift [hunk 4]
```diff
@@ -290,22 +290,41 @@ extension SwiftPackageTool {
...
     struct APIDiff: SwiftCommand {
         static let configuration = CommandConfiguration(
-            commandName: "experimental-api-diff")
+            commandName: "experimental-api-diff",
+            abstract: "Diagnose API-breaking changes to Swift modules in a package",
+            discussion: """
+            The experimental-api-diff command can be used to compare the Swift API of \
+            a package to a baseline revision, diagnosing any breaking changes which have \
+            been introduced. By default, it compares every Swift module from the baseline \
+            revision which is part of a library product. If this behavior is undesirable, \
```
Comment (elsh) [inline Sources/Commands/SwiftPackageTool.swift:299]: Please add a comment on the performance, for example, "This can result in a slow performance if there are a large number of modules. If this behavior is ..."

File: Sources/Commands/SwiftPackageTool.swift [hunk 5]
```diff
@@ -290,22 +290,41 @@ extension SwiftPackageTool {
...
+            the comparison.
+            """)
 
         @OptionGroup(_hiddenFromHelp: true)
         var swiftOptions: SwiftToolOptions
 
         @Argument(help: "The baseline treeish to compare to (e.g. a commit hash, branch name, tag, etc.)")
         var treeish: String
 
+        @Option(name: .customLong("product"), help: "One or more products to include in the diff")
```
Comment (elsh) [inline Sources/Commands/SwiftPackageTool.swift:310]: Let's use a plural "products" since there can be multiple.

File: Sources/Commands/SwiftPackageTool.swift [hunk 6]
```diff
@@ -290,22 +290,41 @@ extension SwiftPackageTool {
...
+        var products: [String] = []
+
+        @Option(name: .customLong("target"), help: "One or more targets to include in the diff")
```
Comment (elsh) [inline Sources/Commands/SwiftPackageTool.swift:313]: Ditto, "targets"

File: Sources/Commands/APIDigester.swift [hunk 7]
```diff
@@ -116,77 +133,137 @@ struct APIDigesterBaselineDumper {
...
-        try apiDigesterTool.dumpSDKJSON(
-            at: sdkJSON,
-            modules: graph.apiDigesterModules,
-            additionalArgs: buildOp.buildPlan!.createAPIToolCommonArgs(includeLibrarySearchPaths: false)
-        )
+        try localFileSystem.createDirectory(baselineDir, recursive: true)
+        let group = DispatchGroup()
+        let errors = ThreadSafeArrayStore<Swift.Error>()
+        for module in modulesToDiff {
+            DispatchQueue.sharedConcurrent.async(group: group) {
```
Comment (elsh) [inline Sources/Commands/APIDigester.swift:143]: Thanks for adding this. Could you add support to control concurrency as well (DispatchSemaphore)? Will need to add a "-j" input option to the cmdline. This can be a separate PR though.

**Comment by elsh** [comment]
> @elsh Sure, I've added support for filtering the comparison to a subset of the package. I chose `--product` and `--target` because I thought those would be more intuitive when working on a package, but I'm open to switching to `--module` or `--module-list` if you think that would be better. I've also updated the help text and parallelized the api-digester invocations.

Thanks for making the changes. The flag naming makes sense. Left a few comments.

File: Sources/Commands/SwiftPackageTool.swift (hunk 4 above)
Comment (owenv) [reply Sources/Commands/SwiftPackageTool.swift]: 👍 I reworded this to add the extra note. I also ran a couple of benchmarks locally, and it looks like performance probably won't be a major concern for any package with < 100-200 targets. I've been primarily testing with swift-nio, which has about a dozen, and the diff operation takes under 2 seconds using a release w/ asserts frontend

File: Sources/Commands/SwiftPackageTool.swift (hunk 5 above)
Comment (owenv) [reply Sources/Commands/SwiftPackageTool.swift]: 👍 changed, I also switched the parsing strategy so these can be passed as `--products P1 P2` instead of `--product P1 --product P2`.

File: Sources/Commands/APIDigester.swift (hunk 7 above)
Comment (owenv) [reply Sources/Commands/APIDigester.swift:145]: I ended up making this change in this PR since the command already accepted `-j` to control the parallelism of the build operation. Now both stages should respect the option consistently

**Comment by owenv** [comment]
@elsh thanks for the review!

**Comment by owenv** [comment]
@swift-ci please smoke test

File: Sources/Commands/SwiftPackageTool.swift [hunk 8]
```diff
@@ -290,22 +290,44 @@ extension SwiftPackageTool {
...
-            commandName: "experimental-api-diff")
+            commandName: "experimental-api-diff",
+            abstract: "Diagnose API-breaking changes to Swift modules in a package",
+            discussion: """
+            The experimental-api-diff command can be used to compare the Swift API of \
+            a package to a baseline revision, diagnosing any breaking changes which have \
+            been introduced. By default, it compares every Swift module from the baseline \
+            revision which is part of a library product. For packages with many targets, this \
+            behavior may be undesirable as the comparison can be slow. \
+            The `--product` and `--target` options may be used to restrict the scope of \
```
Comment (elsh) [inline Sources/Commands/SwiftPackageTool.swift:301]: Update these flags as well --products and --targets.

File: Sources/Commands/APIDigester.swift (hunk 7 above)
Comment (elsh) [reply Sources/Commands/APIDigester.swift:145]: Thanks for adding semaphore.

**Review by elsh — APPROVED** [review]

**Comment by owenv** [comment]
@swift-ci please smoke test
